### PR TITLE
Review 01

### PR DESCRIPTION
**Pull request by sondrele**
# WIP

### CODE REVIEW

File: background/rust.tex [hunk 1]
```diff
@@ -16,13 +16,64 @@ \subsection{Language Features}
...
+\end{listing}
+
+An important part of variable bindings is Rust's way to automatically infer the data type of the variables.
+We can see from the example that \texttt{a} is defined to be a 32-bit integer with the value 5, and because the variable \texttt{b} later gets assigned to \texttt{a}'s value, Rust will automatically infer \texttt{b} to be of the same type.
 
 \subsubsection{Sum types - Enum}
 
-The \textbf{enum} type of Rust is a powerful construct.
-The type is based on sum types and can hold tagged data in addition to named values as enumerated types as found in C.
+A sum type, also known as a tagged union, or an \texttt{enum} in Rust, is a data structure that is used to hold only one out of a small set of possible values.
```
Comment (havardh) [inline background/rust.tex:49]: I don't think the `only` is necessary here.

File: background/rust.tex [hunk 2]
```diff
@@ -16,13 +16,64 @@ \subsection{Language Features}
...
+      None,
+    }
+  \end{minted}
+  \caption{Definition of Option}
+  \label{lst:rust:option}
+\end{listing}
+
+Every \texttt{Option} variable is an \texttt{enum} that can either have the algebraic value \texttt{Some} or the \emph{named value} \texttt{None}.
+If the variable is is \texttt{Some}, the definition also says that it needs to contain any another type \texttt{T}.
+This is also an example of generics in Rust, \texttt{T} can be anything from an integer to a string.
```
Comment (havardh) [inline background/rust.tex]: `... can be anything e.g an integer, a string or an user defined type.`

Comment (havardh) [inline background/rust.tex]: As to not imply that the type is either an integer or a string.

File: background/rust.tex [hunk 3]
```diff
@@ -16,13 +16,64 @@ \subsection{Language Features}
...
+The \texttt{Option} type is analogous with a normal C pointer value, which either has the value 0, or \texttt{null} (analogous to \texttt{None}), or a value that that points to a structure at a place in memory (analogous to \texttt{Some}).
+The usage of \texttt{enum} types described in this section become very expressive in combination with \emph{match expressions}, as discussed in \autoref{sub:pattern_matching}.
```
Comment (havardh) [inline background/rust.tex:71]: I suggest more emphasis on the safety of using `Option`compared to a nullptr. 

Perhaps mentioning that the only safe way of observing the value contained inside the `Option` is to use a `match` expression. And that this guarantees that the variable is `null checked` thus preventing nullpointer dereferences. We should also probably mention the downsides vs a pointer as an `Option` is not a zero cost abstraction.

File: background/rust.tex [hunk 4]
```diff
@@ -35,33 +86,56 @@ \subsubsection{Sum types - Enum}
...
 \end{listing}
 
-\subsubsection{Pattern Matching and Destruction}
+\subsubsection{Pattern Matching and Destructuring}
+\label{sub:pattern_matching}
+
+Pattern matching is a powerful language construct of Rust that resembles the \texttt{switch} statement from C.
+It is possible to match against any value or variable in the language, and branch to different blocks of code based on the outcome of the \texttt{match}.
+Another important feature of the \texttt{match} statement is that it needs to be \emph{exhaustive}, this means that all possible outcomes of a \texttt{match} need to be considered, otherwise Rust will make it a compile-time error.
+This is contrary to C, where only primitive values can be used in a \texttt{switch} statement, and there is no requirement to cover all switch cases, e.g. with the \texttt{default} keyword.
```
Comment (havardh) [inline background/rust.tex]: Reword to this: `... e.g. the \texttt{default} keyword is not required.` ?

But this kind of reduces the Rust mechanism as it is a compile time check for validating that you have covered all possible values. Perhaps we should point to an example of refactoring an library by adding a variant to an `enum` causing the compiler to ensure that you add cases to all your matches ?

File: .gitignore [hunk 5]
```diff
@@ -22,4 +22,4 @@
 *.fls
 *.out.ps
 *.pyg
-.standalone.tex
```
Comment (havardh) [inline .gitignore:25]: Still need this one.

File: background/background.tex [hunk 6]
```diff
@@ -3,6 +3,10 @@
 \chapter{Background}
 \label{chap:background}
 
+The section contains all the background information that was deemed necessary in order to understand the contents of the next chapters.
+We start by presenting the hardware and microcontrollers and its software suite that we utilize in this project, before we move on to present the Rust programming language and its package manager.
+We assume that the reader is fairly known with common programming languages like Java and C++, in order to understand some of the fundamental differences with memory management between these languages and Rust.
```
Comment (havardh) [inline background/background.tex]: Perhaps add C to the language mix.

File: background/objectoriented.tex [hunk 7]
```diff
@@ -2,11 +2,13 @@ \section{Object Oriented Embedded Programming}
 \label{sec:back:oo}
```
Comment (havardh) [inline background/objectoriented.tex:2]: This entire file is being rewritten. So the changes here can be reverted

File: background/rust.tex [hunk 8]
```diff
@@ -3,26 +3,77 @@
...
-Rust \cite{web:rust_lang} is a new open source systems programming language developed by the Mozilla Foundation \cite{web:mozilla_foundation}.
-It is designed to support concurrency and parallelism and take full utilization of the underlying hardware.
-Rust makes strong guarantees about isolation and memory safety, while still focusing on performance.
-
+Rust \cite{web:rust_lang} is a new open source systems programming language developed with backing from the Mozilla Foundation \cite{web:mozilla_foundation}.
 Rust is a strongly and statically typed, multi-paradigm programming language.
-The language borrows many traits and features from other, already existing languages.
-The following subsections discusses some of the more prominent language features, that all work together in order to achieves Rust's main goals; mainly full memory safety without sacrificing performance.
-\todo{I've just been throwing out ideas and statements that are probably not very true.}
+The language borrows many traits and features from other programming languages, some of which is described in the following sections.
```
Comment (havardh) [inline background/rust.tex]: Its confusing to use `traits` here as its a feature of Rust. Use `constructs` instead.

File: background/rust.tex [hunk 9]
```diff
@@ -331,80 +415,87 @@ \subsubsection{Lifetimes}
...
 \subsection{Unsafe Code} % (fold)
 \label{ssub:unsafe_code}
 
-The strong type system along with lifetime inference and borrow-checking help Rust become a safe language.
-The language provides safe, zero-cost abstractions over the machine it is running on, but it is not always possible to guarantee full safety without sacrificing performance, or vice versa.
-Achieving good performance is often a synonym with direct memory access and pointer manipulation, which in itself is not considered to be safe operations.
-\todo{At least by Rust. Depends on who considers it.}
+Rust's strong type system and static guarantees about memory safety goes a long way of verifying the safety of programs, but for certain programs the restrictions set by the compiler might be too conservative.
+There exists many programs that are indeed safe, but still not possible for the compiler to verify.
+Examples of code that the compiler cannot verify is code that utilize \emph{raw pointers}, i.e. pointers to mutable data (like a C pointer), code that call directly \emph{into} external C libraries through Rust's \gls{ffi}, or code that can be called \emph{from} \gls{ffi}.
```
Comment (havardh) [inline background/rust.tex]: Remove `i.e. pointers to mutable data` as this is not a definition or sole example of raw pointers. 
A raw pointer is a pointer that is not considered by the ownership analysis.

Comment (havardh) [inline background/rust.tex]: My suggestion to this sentence is: `Examples of code that the compiler cannot verify is code that utilizes \emph{raw pointers}, pointers which are not considered by ownership analysis and code that calls a compiler intrinsic like \emph{volatile_load}.`

File: background/rust.tex [hunk 10]
```diff
@@ -331,80 +415,87 @@ \subsubsection{Lifetimes}
...
+  \hline
+  \raggedleft{Access and update static mutable variables} &
+  Writing and reading to and from global state can lead to race conditions \\
+
+  \raggedleft{Dereference raw pointers} &
+  Raw pointers introduces memory aliasing to mutable data \\
 
-% subsubsection unsafe_code (end)
+  \raggedleft{Call unsafe functions \\ (e.g. over \gls{ffi})} &
+  Rust can make no guarantees about the safety of \texttt{unsafe functions}. If a \gls{ffi} function is called we are leaving Rust and entering another language domain, it is not possible for Rust to verify the safety of such code. \\
```
Comment (havardh) [inline background/rust.tex]: Maybe tone done the emphasis on ffi's, compiler intrinsics like `volatile_load` are also considered unsafe.

Comment (havardh) [inline background/rust.tex]: Although I like to have this exact explaination. So maybe include an instrinsic in the previous example, see above

File: background/rust.tex [hunk 11]
```diff
@@ -331,80 +415,87 @@ \subsubsection{Lifetimes}
...
-The language provides safe, zero-cost abstractions over the machine it is running on, but it is not always possible to guarantee full safety without sacrificing performance, or vice versa.
-Achieving good performance is often a synonym with direct memory access and pointer manipulation, which in itself is not considered to be safe operations.
-\todo{At least by Rust. Depends on who considers it.}
+Rust's strong type system and static guarantees about memory safety goes a long way of verifying the safety of programs, but for certain programs the restrictions set by the compiler might be too conservative.
+There exists many programs that are indeed safe, but still not possible for the compiler to verify.
+Examples of code that the compiler cannot verify is code that utilize \emph{raw pointers}, i.e. pointers to mutable data (like a C pointer), code that call directly \emph{into} external C libraries through Rust's \gls{ffi}, or code that can be called \emph{from} \gls{ffi}.
+
+The ability to make new abstractions that optimize away to essentially nothing is an important feature of Rust.
```
Comment (havardh) [inline background/rust.tex:475]: Suggest to rephrase to `The ability to make abstractions that optimize away with no overhead is an important design goal of Rust, called Zero Cost Abstraction.`

File: background/rust.tex [hunk 12]
```diff
@@ -331,80 +415,87 @@ \subsubsection{Lifetimes}
...
+Sometimes this means that the underlying code needs direct memory access and the ability to dereference raw pointers or call in to external libraries.
+For these tasks, Rust provides a new keyword, \texttt{unsafe}.
```
Comment (havardh) [inline background/rust.tex]: `For these tasks, Rust provides the keyword \texttt{unsafe}.`

File: background/rust.tex [hunk 13]
```diff
@@ -331,80 +415,87 @@ \subsubsection{Lifetimes}
...
+It is important to note that this \texttt{unsafe} keyword is introduced in order to \emph{keep} Rust's semantics about memory safety, when the compiler can no longer guarantee it, it is up to the programmer to tell the compiler that the code inside the \texttt{unsafe} block is indeed safe.
```
Comment (havardh) [inline background/rust.tex]: Should we reference the discussions where people misunderstands this fact?

File: background/rust.tex [hunk 14]
```diff
@@ -331,80 +415,87 @@ \subsubsection{Lifetimes}
...
+\autoref{tab:reasons_for_unsafe} summarizes the three things that are possible to do in an \texttt{unsafe} block, and why the operations are considered unsafe.
+It is important to notice that Rust's borrow checker is still active across the unsafe code, and that it will still issue errors during compilation if the rules about ownership is not withheld.
```
Comment (havardh) [inline background/rust.tex]: `are not withheld`. 

Can you explain this a bit further? This is the fact that the sentence `the borrow checker is disabled in unsafe section` is false, right?

File: background/rust.tex [hunk 15]
```diff
@@ -455,7 +546,7 @@ \subsection{Cargo}
 \subsubsection{Project Structure}
 
 Every crate, or package, built by \texttt{Cargo} requires a \texttt{Cargo.toml} file to be present in the root directory.
-This file is interpreted by \texttt{Cargo} and is used to determine the name of library and executables to be built in the package and what other packages that it depends on.
+This file is interpreted by \texttt{Cargo} and is used to determine the name of the library and executables to be built in the package and what other packages that it depends on.
 It also includes information that tells \texttt{Cargo} how the package can be compiled for different target architectures, and it is used to define different \textit{features} of a package - a way to conditionally compile certain parts of the code present in the library.
 
 \begin{listing}
```
Comment (havardh) [inline background/rust.tex:564]: Didn't we agree to simplify this structure?

I suggest something like this as the first example:

```
\dirtree{%
.1 hello\_world.
.2 Cargo.lock.
.2 src.
.3 main.rs.
}
```

File: background/rust.tex (hunk 11 above)
Comment (sondrele) [inline background/rust.tex:475]: This is kind of mentioned in the Zero Cost Abstraction chapter already. Maybe remove the sentence?

File: background/rust.tex (hunk 10 above)
Comment (sondrele) [inline background/rust.tex]: Should probably make the table into its own paragraphs describing the three points a little more thoroughly.

File: background/rust.tex (hunk 11 above)
Comment (havardh) [inline background/rust.tex:475]: No I think it is important to emphasis that unsafe if there to enable ZCA.

**Pull request merged**